test/test_fori_loop_with_while_loop_simple_add_dispatch_in_torch.py

```diff
@@ -19,88 +29,311 @@ def _fake_while_loop(cond_fn, body_fn, operands):
+
+    self.assertTrue(torch.all(torch.eq(res, expected)))
+
+  def test_while_loop_tpu_MNIST_inside_loop(self):
+    xm.mark_step()
     device = xm.xla_device()
+    torch.set_grad_enabled(False)
 
-    def cond_fn(init, limit_value):
-      return limit_value[0] <= init[0]
+    n_epochs = 3
+    batch_size_train = 8
+    batch_size_test = 10
+    learning_rate = 0.01
+    momentum = 0.5
+    log_interval = 10
+    random_seed = 1
+    torch.backends.cudnn.enabled = False
+    torch.manual_seed(random_seed)
 
-    def body_fn(init, limit_value):
-      one_value = torch.ones(1, dtype=torch.int32, device=device)
-      two_value = limit_value.clone()
-      return (torch.sub(torch.sub(init, one_value), one_value), two_value)
+    ### load data
+    test_loader = xu.SampleGenerator(
+    data=(torch.zeros(8, 1, 28,28), torch.zeros(8, dtype=torch.int64)),
+    sample_count=1000 // 8 // xm.xrt_world_size())
 
-    init = torch.tensor([10], dtype=torch.int32, device=device)
-    limit_value = torch.tensor([0], dtype=torch.int32, device=device)
-    res = while_loop(cond_fn, body_fn, (init, limit_value))
-    expected = _fake_while_loop(cond_fn, body_fn, (init, limit_value))
-    self.assertEqual(expected, res)
+    class MNIST(torch.nn.Module):
+      def __init__(self):
+        super().__init__()
+        self.conv1 = torch.nn.Conv2d(1, 10, kernel_size=5, stride=1, padding=2)
+        self.bn1 = torch.nn.BatchNorm2d(10)
+        self.conv2 = torch.nn.Conv2d(10, 20, kernel_size=5)
+        self.bn2 = torch.nn.BatchNorm2d(20)
+        self.fc1 = torch.nn.Linear(500, 50)
+        self.fc2 = torch.nn.Linear(50, 10)
 
-  def test_fori_loop_tpu_addition(self):
+      def forward(self, iteri, x, y):
+        def cond_fn(iteri, x, y):
+          return iteri > 0
+
+        def body_fn(iteri, x, y):
+          y = F.relu(F.max_pool2d(self.conv1(x), 2))
+          # z = self.bn1(z) # torch.while_loop's body_fn might be modifying the input!
+          y = F.relu(F.max_pool2d(self.conv2(y), 2))
+          # y = self.bn2(y)
+          y = torch.flatten(y, 1)
+          y = F.relu(self.fc1(y))
+          y = self.fc2(y)
+
+          return iteri - 1, x.clone(), F.log_softmax(y, dim=1)
+
+        return while_loop(cond_fn, body_fn, (iteri, x, y))
+
+      def forward_compare(self, iteri, x, y):
+        y = F.relu(F.max_pool2d(self.conv1(x), 2))
+        # z = self.bn1(z) # torch.while_loop's body_fn might be modifying the input!
+        y = F.relu(F.max_pool2d(self.conv2(y), 2))
+        # y = self.bn2(y)
+        y = torch.flatten(y, 1)
+        y = F.relu(self.fc1(y))
+        y = self.fc2(y)
+        return iteri - 1, x.clone(), F.log_softmax(y, dim=1)
+
+    mnist = MNIST()
+    mnist.to(device)
+    bs=16
+    l_in_0 = torch.randn(bs, 1, 28, 28, dtype=torch.float32, device=device)
+    l_out = torch.randn(bs, 10, dtype=torch.float32, device=device)
+    iteri = torch.tensor(3, dtype=torch.int64, device=device)
+    _, _, res = mnist(iteri, l_in_0, l_out)
```

Comment:
can we init the `l_out ` inside the mnist forward and returns it? This way the only additional output we provide is the  count.

Reply:
thanks, good idea, due to hard to gain result before finish one run, so current plan would be: 
run body_fn with `pure python while`, get result, then use this result as `l_out`

before implementation, did quick benchmark for this idea with [code](https://gist.github.com/ManfeiBai/482deb202488009a2dea3326b7e4dea6):
```
- iteri = 3: 
  while_loop(3.675 s), pure while(3.107 s), pure_while_then_while_loop_plan(3.724 s)
- iteri = 10: 
  while_loop(3.670s), pure while(3.119s), pure_while_then_while_loop_plan(3.932s)
- iteri = 10: 
  while_loop(3.744 s), pure while(3.110 s), pure_while_then_while_loop_plan(3.749 s)
- iteri = 10: 
  while_loop(3.878 s), pure while(7.053 s), pure_while_then_while_loop_plan(3.809 s)
```
so this plan looks like worth to try